include/oneapi/dpl/pstl/onedpl_config.h

```diff
@@ -49,11 +49,29 @@
 #    define _ONEDPL_PREDEFINED_POLICIES 1
 #endif
 
-#if ONEDPL_USE_TBB_BACKEND || (!defined(ONEDPL_USE_TBB_BACKEND) && !ONEDPL_USE_OPENMP_BACKEND)
+// Check availability of parallel backends
+#if __has_include(<tbb/tbb.h>)
+#    define _ONEDPL_TBB_AVAILABLE 1
+#endif
+#if ONEDPL_USE_TBB_BACKEND && !_ONEDPL_TBB_AVAILABLE
+#    error "TBB backend is selected, but TBB is not found"
+#endif
+// Check OpenMP availability only during host code compilation
+#if !defined(__SYCL_DEVICE_ONLY__)
```

Comment:
With Intel® oneAPI DPC++/C++ Compiler and device code compilation (`-fsycl`), the code is preprocessed twice: for host and device.
`__SYCL_DEVICE_ONLY__` is set during the device compilation pass, when `_OPENMP` is not defined regardless of the flags set (e.g. `-fiopenmp` or even `-fopenmp-targets=spir64`)

Reply:
To be more precise, `-fopenmp-targets` option one more preprocessor pass.

Reply:
I performed an experiment with the following code and different compilers/options:

```c++
#if defined(_OPENMP) && defined(__SYCL_DEVICE_ONLY__)
#   warning "1 && 1"
#endif
#if defined(_OPENMP) && !defined(__SYCL_DEVICE_ONLY__)
#   warning "1 && 0"
#endif
#if !defined(_OPENMP) && defined(__SYCL_DEVICE_ONLY__)
#   warning "0 && 1"
#endif
#if !defined(_OPENMP) && !defined(__SYCL_DEVICE_ONLY__)
#   warning "0 && 0"
#endif

int main() {}
```

Intel® oneAPI DPC++/C++ Compiler:
|                                          | 00 | 01 | 10 | 11 |
|------------------------------------------|----|----|----|----|
| no options                               | +  |    |    |    |
| -fsycl                                   | +  | +  |    |    |
| -fiopenmp                                |    |    | +  |    |
| -fsycl -fiopenmp                         |    | +  | +  |    |
| -fiopenmp -fopenmp-targets=spir64        |    |    | ++ |    |
| -fsycl -fiopenmp -fopenmp-targets=spir64 |    | +  | ++ |    |

Clang 15.0.0:
|                                         | 00 | 01 | 10 | 11 |
|-----------------------------------------|----|----|----|----|
| no options                              | +  |    |    |    |
| -fsycl                                  |    | +  |    |    |
| -fopenmp                                |    |    | +  |    |
| -fsycl -fopenmp                         |    |    |    | +  |
| -fopenmp -fopenmp-targets=x86_64        |    |    | ++ |    |
| -fsycl -fopenmp -fopenmp-targets=x86_64 |    |    |    | ++ |




 - The line below means that there have been two warnings (`0 && 0` and `0 && 1`) and thus two preprocessor passes:

|                                          | 00 | 01 | 10 | 11 |
|------------------------------------------|----|----|----|----|
| -fsycl                                   | +  | +  |    |    |

- `++` means two the same warnings appeared (two preprocessor passes).

Reply:
Here is the conclusion.  There may be multiple preprocessor passes depending on a compiler and options used. They are different and checking only `!__SYCL_DEVICE_ONLY__` is not enough to verify that the current pass is for host.

@rarutyun, could you suggest a better (and more reliable option)? If there is no such an option, I lean toward removing notifications/errors about backends selected by a user, when they are not available.

Reply:
The purpose is not to find whether a pass is for host, but to avoid checking for OpenMP presence on a device. I.e. we should avoid false negative errors when OpenMP was really specified - that would be "00" or "01" in the rows with `-fiopenmp`. And we can see that indeed there is no "00" but there might be "01" caused by the combination of two options. So limiting the check to only the "00" case, as it is done in the current implementation, seems practically fine to me.